### PR TITLE
[2] Fix/237

### PR DESCRIPTION
**Pull request by fgrunewald**
Allows parsing and handling of volumes and residue templates. However, I don't like the mess of different processors setting hidden attributes. I think I'll get rid of the volumes and templates attribute of topology and meta-molecule and instead hand them down as processor arguments.

### CODE REVIEW

File: polyply/src/build_file_parser.py [hunk 1]
```diff
@@ -125,11 +128,54 @@ def _persistence_length(self, line, lineno=0):
         specs = PersistenceSpecs(*[model, persistence_length, start, stop, self.current_molidxs])
         self.topology.persistences.append(specs)
 
+    @SectionLineParser.section_parser('template')
+    def _template(self, line, lineno=0):
+        """
+        Parses the lines in the '[template]'
+        directive and stores it.
+        """
+        name = line.split()[1]
```
Comment (pckroon) [inline polyply/src/build_file_parser.py:137]: What's the first element? Maybe add an example in the docstring?

File: polyply/src/build_file_parser.py [hunk 2]
```diff
@@ -125,11 +128,54 @@ def _persistence_length(self, line, lineno=0):
...
+        """
+        resname, volume = line.split()
+        self.topology.volumes[resname] = float(volume)
+
     def finalize(self, lineno=0):
         """
         Tag each molecule node with the chirality and build options
         if that molecule is mentioned in the build file by name.
         """
+        print("go here")
```
Comment (pckroon) [inline polyply/src/build_file_parser.py:178]: leftover

File: polyply/src/generate_templates.py [hunk 3]
```diff
@@ -302,9 +299,12 @@ def __init__(self, topology, max_opt, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.templates = {}
         self.resnames = []
-        self.volumes = {}
         self.max_opt = max_opt
         self.topology = topology
+        if "volumes" in self.topology:
+            self.volumes = self.topology.volumes
+        else:
+            self.volumes = {}
```
Comment (pckroon) [inline polyply/src/generate_templates.py:307]: Is topology a dict? If so:
```suggestion
        self.volumes = self.topology.get("volumes", {})
```

File: polyply/src/generate_templates.py [hunk 4]
```diff
@@ -325,10 +325,8 @@ def _gen_templates(self, meta_molecule):
         """
         resnames = set(nx.get_node_attributes(meta_molecule.molecule,
                                               "resname").values())
-        templates = {}
-        volumes = {}
         for resname in resnames:
-            if not resname in self.resnames:
+            if not resname in self.resnames and resname not in meta_molecule.templates:
```
Comment (pckroon) [inline polyply/src/generate_templates.py:329]: I would suggest changing the order of the if/elif/else around a bit. First the simple cases, then the complex ones.

File: polyply/src/generate_templates.py [hunk 5]
```diff
@@ -349,19 +352,24 @@ def _gen_templates(self, meta_molecule):
...
                 coords = map_from_CoG(coords)
                 self.templates[resname] = coords
+            elif resname in meta_molecule.templates:
+                coords = meta_molecule.templates[resname]
+                coords = map_from_CoG(coords)
+                self.templates[resname] = coords
+            else:
+                coords = None
 
-        return templates, volumes
```
Comment (pckroon) [inline polyply/src/generate_templates.py:357]: Is it intentional this method doesn't return anything anymore?

File: polyply/tests/test_build_file_parser.py [hunk 6]
```diff
@@ -264,3 +265,81 @@ def test_persistence_parsers(test_system, lines, expected):
...
+                                                  test_system.molecules,
+                                                  test_system)
+    for mol in test_system.molecules:
+        assert len(mol.templates) == len(names)
+        for idx, name in enumerate(names):
+            template = mol.templates[name]
+            assert edges[idx] == list(template.edges)
+            for node_pos in positions[idx]:
+                node = node_pos[0]
+                assert all(np.array(node_pos[1:], dtype=float) == template.nodes[node]["position"])
```
Comment (pckroon) [inline polyply/tests/test_build_file_parser.py:324]:
```suggestion
                assert np.all(np.array(node_pos[1:], dtype=float) == template.nodes[node]["position"])
```

**Review by pckroon — CHANGES_REQUESTED** [review]

File: polyply/src/generate_templates.py (hunk 5 above)
Comment (fgrunewald) [reply polyply/src/generate_templates.py:357]: yes in fact the function updates some class attributes

File: polyply/src/generate_templates.py [hunk 7]
```diff
@@ -180,9 +180,8 @@ def compute_volume(molecule, block, coords, nonbond_params, treshold=1e-18):
         rad = float(nonbond_params[frozenset([atom_key, atom_key])]["nb1"])
         diff = coord - res_center_of_geometry
         if np.linalg.norm(diff) < treshold:
-           continue
-        else:
-           geom_vects[idx, :] = diff + u_vect(diff) * rad
+            continue
+        geom_vects[idx, :] = diff + u_vect(diff) * rad
         idx += 1
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply polyply/src/generate_templates.py]: agreed

**Review by jan-stevens — APPROVED** [review]

**Pull request merged**